src/field_theory/separable.lean

```diff
@@ -221,6 +227,99 @@ lemma expand_injective {n : ℕ} (hn : 0 < n) :
   exact h',
 end
 
+lemma is_unit_of_self_mul_dvd_separable {p q : polynomial R}
+  (hp : p.separable) (hq : q * q ∣ p) : is_unit q :=
+begin
+  obtain ⟨p, rfl⟩ := hq,
+  apply is_coprime_self.mp,
+  have : is_coprime (q * (q * p)) (q * (q.derivative * p + q.derivative * p + q * p.derivative)),
+  { simp only [← mul_assoc, mul_add],
+    convert hp,
+    rw [derivative_mul, derivative_mul],
+    ring },
+  exact is_coprime.of_mul_right_left (is_coprime.of_mul_left_left this)
+end
+/-- The opposite of `expand`: sends `∑ aₙ xⁿᵖ` to `∑ aₙ xⁿ`. -/
+noncomputable def contract (p : ℕ) (f : polynomial R) : polynomial R :=
+∑ n in range (f.nat_degree + 1), monomial n (f.coeff (n * p))
+
+theorem coeff_contract {p : ℕ} (hp : p ≠ 0) (f : polynomial R) (n : ℕ) :
+  (contract p f).coeff n = f.coeff (n * p) :=
+begin
+  simp only [contract, coeff_monomial, sum_ite_eq', finset_sum_coeff, mem_range, not_lt,
+    ite_eq_left_iff],
+  assume hn,
+  apply (coeff_eq_zero_of_nat_degree_lt _).symm,
+  calc f.nat_degree < f.nat_degree + 1 : nat.lt_succ_self _
+    ... ≤ n * 1 : by simpa only [mul_one] using hn
+    ... ≤ n * p : mul_le_mul_of_nonneg_left (show 1 ≤ p, from hp.bot_lt) (zero_le n)
+end
+
+section char_p
+
+variable [HF : char_p R p]
+include HF
+
+theorem expand_contract [no_zero_divisors R] {f : polynomial R} (hf : f.derivative = 0)
+  (hp : p ≠ 0) : expand R p (contract p f) = f :=
+begin
+  ext n,
+  rw [coeff_expand hp.bot_lt, coeff_contract hp],
+  split_ifs with h,
+  { rw nat.div_mul_cancel h },
+  { cases n,
+    { exact absurd (dvd_zero p) h },
+    have := coeff_derivative f n,
+    rw [hf, coeff_zero, zero_eq_mul] at this,
+    cases this,
+    { rw this },
+    rw [← nat.cast_succ, char_p.cast_eq_zero_iff R p] at this,
+    exact absurd this h }
+end
+
+variable [hp : fact p.prime]
+include hp
+
+theorem expand_char (f : polynomial R) : map (frobenius R p) (expand R p f) = f ^ p :=
```

Comment:
Shouldn't this be called `map_frobenius_expand`? Or are you just moving this lemma?

Reply:
all the lemmas here (apart from `separable_of_subsingleton`) were just moved; I'm still happy to rename this, though; it seems a better name

Reply:
I think we should go through the whole file in a separate PR since there's some suboptimal naming throughout e.g. the pattern `whatever_of_separable` to `separable.whatever`.